### PR TITLE
Fix internal compiler error when assigning function marked pure or view to a variable of function type

### LINKED ISSUE #14999: SMTChecker fails to handle arrays of pure/view functions
## Description

The following code triggers an `SMT logic error`. This appears to be due to `solc` treating `function () external returns (uint256)_array_tuple` as distinct from `function () view external returns (uint256)_array_tuple`. However, the code compiles successfully, indicating that `solc`'s type checking deems it valid.

```solidity
contract C {

  function() external returns(uint)[1] a;

  function b() external view returns(uint) {
    return 1;
  }

  function test() public returns(uint) {
    a = [this.b];
    return a[0]();
  }
}
```

Compile: `solc  --model-checker-engine all a.sol`

Output:

```
SMT logic error:
/solidity/libsmtutil/Z3Interface.cpp(270): Throw in function z3::expr solidity::smtutil::Z3Interface::toZ3Expr(const solidity::smtutil::Expression&)
Dynamic exception type: boost::wrapexcept<solidity::smtutil::SMTLogicError>
std::exception::what: Sorts |function () external returns (uint256)_array_tuple| and |function () view external returns (uint256)_array_tuple| are incompatible
[solidity::util::tag_comment*] = Sorts |function () external returns (uint256)_array_tuple| and |function () view external returns (uint256)_array_tuple| are incompatible
```

The same issue occurs if function `b` is defined as a `pure` function.

## Environment

- Compiler version: 0.8.25

### PR DESCRIPTION
**Pull request by pgebal**
Closes https://github.com/ethereum/solidity/issues/14999

### CODE REVIEW

File: Changelog.md [hunk 1]
```diff
@@ -20,6 +20,7 @@ Compiler Features:
 Bugfixes:
  * SMTChecker: Fix error that reports invalid number of verified checks for BMC and CHC engines.
  * SMTChecker: Fix formatting of unary minus expressions in invariants.
+ * SMTChecker: Fix internal compiler error when assigning function marked pure or view to a function type variable.
```
Comment (matheusaaguiar) [inline Changelog.md:23]:
```suggestion
 * SMTChecker: Fix SMT logic error when assigning function marked pure or view to a function type variable.
```

Shouldn't it be `SMT logic error`? I read https://github.com/ethereum/solidity/issues/14999 and from what I understood, the compiler considers both types the same, but the smt was seeing them as different, right?

File: libsolidity/formal/SymbolicTypes.h [hunk 2]
```diff
@@ -86,4 +86,5 @@ std::optional<smtutil::Expression> symbolicTypeConversion(frontend::Type const*
 smtutil::Expression member(smtutil::Expression const& _tuple, std::string const& _member);
 smtutil::Expression assignMember(smtutil::Expression const _tuple, std::map<std::string, smtutil::Expression> const& _values);
 
+void removePureAndViewModifiers(std::string& typeName);
```
Comment (matheusaaguiar) [inline libsolidity/formal/SymbolicTypes.h:89]:
```suggestion
void removePureAndViewModifiers(std::string& _typeName);
```

nitpick

**Review by matheusaaguiar — COMMENTED** [review]
Looks good overall. Just have a question about the Changelog and a small nitpick.

**Comment by blishko** [comment]
This fixes this particular problem, but I would like to find a more general solution, one that would also help with #15308. We are still looking into some options there.

**Comment by pgebal** [comment]
Better fix was merged here: https://github.com/ethereum/solidity/pull/15322

**Pull request closed without merging**